Accept arrays in parse_config's method_missing (#5252)
chef/inspec: docs-chef-io/content/inspec/resources/parse_config_file.md

- May run a command to `stdout`, and then run the test against that output
- May use options to define how configuration data is to be parsed

> ### Option names containing dots

Reviewer: ```suggestion
### Option Names Containing Periods
```

Website formatting prefers to have the first letter of each word in a header capitalized.

> A possibly unexpected behaviour may occur when attempting to access option names containing dots with `its()`. A known behaviour, there are two ways to work around it:

Reviewer: ```suggestion
A possible behavior may occur when attempting to access option names containing periods with `its()`. There are two ways to work around it:
```

* Access the option by using the `params` attribute of the returned resource object

> * Since 4.23.9, `its` can be used by passing the option name in a single-element array (cf. examples)

Reviewer: ```suggestion
* Since 4.23.9, `its` can be used by passing the option name in a single-element array. See the `parse config file` examples.
```

Reviewer: After this PR merges, I will insert an internal reference link to the `parse config file` Examples section as a part of a link clean-up project in the docs. So know that this is planned and won't be omitted in the long term.

Author: I bumped the version number but it should be fixed once the PR is actually merged.

    'key = value # comment'
    params['key'] = 'value'

## Matchers

For a full list of available matchers, please visit our [matchers page](/inspec/matchers/).

## Examples

The following examples show how to use this Chef InSpec audit resource.

> ### Test a configuration setting

Reviewer: ```suggestion
### Test A Configuration Setting
```

    describe parse_config_file('/path/to/file.conf') do
     its('PARAM_X') { should eq 'Y' }
    end

> ### Use options, and then test a configuration setting

Reviewer: ```suggestion
### Use Options And Then Test A Configuration Setting
```

    describe parse_config_file('/path/to/file.conf', { multiple_values: true }) do
     its('PARAM_X') { should include 'Y' }
    end

> ### Test a file with an ini-like structure (such as a yum.conf)

Reviewer: ```suggestion
### Test A File With An INI File Structure
```

> 

Reviewer: ```suggestion
`yum.conf` is one example of an INI file structure type. 

```

    describe parse_config_file('/path/to/yum.conf') do
     its('main') { should include('gpgcheck' => '1') }
    end

> ### Test a configuration setting containing dots

Reviewer: @rgeissert When referring to "dots," are you talking about periods (`.`) within a URL, a specific technical file, or something else?

Author: @mjingle It should read: "Test a configuration setting whose name contains dots" (or periods, or full stops, as you prefer :)). As in: `author.name = Raphael` in a configuration file.

Reviewer: @rgeissert Awesome! Thank you for confirming. That will help inform my suggested edits to you, and then we'll get this docs part finished.

> ### Test a configuration setting containing dots

Reviewer: ```suggestion
### Test A Configuration Setting Containing Periods
```

    describe parse_config_file('/etc/sysctl.conf') do

>      its(['kernel.domainname']) { should eq 'example.com' }

Reviewer 2: This is just opinion.

This just feels strange - I think this may be a case of implementation driving UX. If I were a user, and I wanted to do a query containing a dot, even assuming I knew it had to be escaped, the typical thing would be to use a slash, not to enclose the query in brackets.

I agree the need is valid, and here we wish we could hijack `rspec-its` dot-processing facility entirely. So why not do so? In method_missing, why not explicitly search for dots and transform the call as needed to fool `rspec-its`? Then there may be no need to alter the DSL syntax at all - you may be able to have:

```
its('kernel.domainname') { should eq 'example.com' }
```

simply work.

Author: From my understanding of rspec-its' code our method_missing won't ever see the dot due to the code at:
https://github.com/rspec/rspec-its/blob/v1.2.0/lib/rspec/its.rb#L113

I drafted some code to return some magic object from method_missing so that the calls to `send` Just Work.
It fails, however, whenever the hash contains "foo.bar" and "foo.bar.baz" and one wants to get the latter.

If that sounds acceptable I can use that instead of accepting arrays (like other resources do already).

I don't know how it should behave with regards to_sth methods. E.g. only "foo.bar.baz" exists, but the caller wants 'foo.bar' only.
On the plus side, `its('foo.bar.baz.downcase')` just works.

Reviewer 2: Ah, thank you, that code reference cleared up the situation for me. I think your existing approach using the array syntax is an acceptable compromise.

Author: If it's okay for you to leave it as-is then I think this pull request is ready from my POV.
